[Jerome Neo] iP

jerome-neo opened the pull request:
<h1>Duke+</h1>

> "Your mind is for having ideas, not holding them." - David Allen [(source)](https://dansilvestre.com/productivity-quotes/)

### Duke+ helps you organise your life, perfect for rememebering:
<li> Events
<li> Deadlines
<li> Todos

~~CROSS~~ out your list _anywhere_ and _anytime_.

Follow these steps:
1. download [here](https://github.com/jerome-neo/ip/releases/tag/A-Jar).
2. double-click
3. key in your commands
4. organise your life!

Most importantly, it is **FREE**!

Road map:
- [x] Basic features
- [ ] Setting reminders
- [ ] UI improvements

Are you a developer?
Here's how you can help improve the app!
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

jerome-neo committed:
Level-1

jerome-neo committed:
Level 2. Add, List

jerome-neo committed:
Level 3. Mark as Done

jerome-neo committed:
Level 4. ToDos, Events, Deadlines

jerome-neo committed:
fixed [D] for deadline

jerome-neo committed:
Level 5. Handle Errors

jerome-neo committed:
Revert "Level 5. Handle Errors"

This reverts commit 1742c92f89c33a5832ad0a54537dd7d9c6821ade.

jerome-neo committed:
Level 5. Handle Errors

jerome-neo committed:
Level 6. Delete

jerome-neo committed:
Update Level-7

jerome-neo committed:
Add Level-8

jerome-neo committed:
Correct input matching issue in Duke.java

jerome-neo committed:
Merge branch 'branch-Level-7'

jerome-neo committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Save.java
#	src/main/java/Task.java

jerome-neo committed:
Add A-MoreOOP

jerome-neo committed:
Add package duke.task

jerome-neo committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

jerome-neo committed:
Add gradle support

jerome-neo committed:
Add JUnit Tests

jerome-neo committed:
Add additional test cases

jerome-neo committed:
Add comment headers for classes

jerome-neo committed:
Merge branch 'branch-A-JavaDoc'

jerome-neo committed:
Add A-CodingStandard

jerome-neo committed:
Add Level-9

jerome-neo committed:
Merge branch 'branch-A-CodingStandard'

jerome-neo committed:
Merge branch 'branch-Level-9'

jerome-neo committed:
Include Checkstyle

leeyi45 commented on src/main/java/duke/main/Duke.java:
```diff
@@ -0,0 +1,73 @@
+                display.line();
+            } else if (input.startsWith("mark")) {
+                int taskNumber = logic.indexToMark(input);
+                tasks.taskMarkedAtIndex(--taskNumber);
```
I personally don't like using the prefix increment thing, because it's usually terribly confusing (increment then return? return then increment?)

vsamarth replied:
I agree with this. Decrementing it first and then using this would be much more clearer.

leeyi45 commented on src/main/java/duke/store/Storage.java:
```diff
@@ -0,0 +1,93 @@
+     * @param task is the task to be converted into a String to be saved.
+     * @return string that is to be saved.
+     */
+    private static String readerFriendly(Task task) {
```
Methods should use verbs. Maybe `getReaderFriendly()`?

leeyi45 commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,45 @@
+ *      Description: Deadline class that inherits from the Task class.
+ */
+public class Deadline extends Task {
+    protected String by;
```
(personally idm this) but there are people who think that fields (especially non final ones) should always be private

leeyi45 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,99 @@
+     * @param stringDate
+     * @return the LocalDateTime representation of the string.
+     */
+    public static LocalDateTime convertDateTime(String stringDate) {
```
I think you should follow the format of your similarly named method `dateTimeToString` and name this `stringToDateTime`

leeyi45 commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,151 @@
+        Integer j = 0;
+        for (int i = 0; i < tasks.size(); i++) {
+            Task task = this.getTaskAtIndex(i);
+            if (task.getDescription().equals(description)) {
```
Consider using `contains`, or some kind of locale based comparisons

jerome-neo committed:
Add Level-10

vsamarth commented on src/data/duke.txt:
```diff
@@ -0,0 +1,2 @@
+Todo: run     ๑(◕‿◕)๑ COMPLETED! ๑(◕‿◕)๑    
+Deadline: interview internship by 01/01/2023 1300     (｡-_-｡ ) INCOMPLETE ( ｡-_-｡)    git 
```
Perhaps removing this from the repository by adding it to .gitignore would be suitable as it is something specific to each user.

vsamarth commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,99 @@
+ *      @author: Jerome Neo
+ *      Description: Task class that Todo, Event and Deadline inherits from.
+ */
+public class Task {
```
Consider marking this class abstract as there are no instantiations of this class and its an abstract entity that other classes inherit from.

vsamarth commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,54 @@
+ */
+public class Ui {
+    private final String LINE = "____________________________________________________________";
+    private final String LOGO = " ____        _        \n"
```
These can be marked as static as they have fixed value.

vsamarth reviewed (approved):
The project overall is very well structured and the focus on details is excellent. 

However, the only thing I noticed was that the abstraction between tasks and the UI are not maintained properly (there are print statements in the task classes).
This would make changing the User Interface more tricky as you would have change your task classes as well.

jerome-neo committed:
A-FullCommitMessage

Current Duke prints output in the terminal but we want a GUI. There are
several additional methods added several classes to handle
returning a string instead of using System.out.println(). This is
because the GUI needs a String.

Added a GUI to the duke application.

Every class is part of a package to make things neater.

jerome-neo committed:
Fix Jar file

jerome-neo committed:
Add assert statements.

There are no assert statements included in the code. As required by the
individual project, we need to include some assert statements.

Added assert statements to Duke.java and Parser.java.

The assert statement confirms that the else clause is in fact working as
intended.

jerome-neo committed:
Improve code quality

Minor code quality issues found that impedes readability.

Refactoring done to if-else statements.

Increase readability will help make the more prominent path obvious.

jerome-neo committed:
Merge pull request #1 from jerome-neo/branch-A-Assertions

Add branch assertions

jerome-neo committed:
Merge branch 'master' into branch-A-CodeQuality

jerome-neo committed:
Merge pull request #2 from jerome-neo/branch-A-CodeQuality

Add branch-CodeQuality

jerome-neo committed:
Amend error from pull request

Files were not of the correct version (GUI version), which resulted in
command line not working.

Reverted files back and included changes that were made after the
merges.

jerome-neo committed:
Added assert to Duke.java

jerome-neo committed:
Add CI

jerome-neo committed:
More flexibility in searching for items

The current find command can only look for 100% matches including spaces
 in the list of tasks.

FuzzySearch class is added. It includes the Levenshtein distance
algorithm (Levenshtein, 1965).

The algorithm will provide similar matches instead of 100% matches.
The threshold is also easily adjustable based on requirements.

jerome-neo committed:
Merge tag 'A-CI'

jerome-neo committed:
Update .gitignore to ignore src/data/

jerome-neo committed:
no message

jerome-neo committed:
Merge tag 'BCD-Extension'

jerome-neo committed:
Update changes with GUI

GUI is missing in the files. GUI files added back in slowly.

jerome-neo committed:
Improve and customize GUI: Change background and increase padding on
chat boxes

The current GUI of the app looks dull and uninviting. Changing the
background color and increasing the padding on chat boxes can enhance
the overall look and feel of the app.

In this commit, I have made the following changes:

Changed the background color to a more vibrant and appealing one
Increased the padding on chat boxes to make them look more spaced out
and prominent.
These changes will make the app look more modern and engaging, which
will likely increase user engagement and satisfaction.

jerome-neo committed:
Update Storage class to automatically create directory

The Storage class was failing to create a directory automatically when
the jar file was running, resulting in errors when trying to save files
to non-existent directories.

To fix this, the code has been modified to check if the directory exists
before creating a file. If the directory does not exist, the code will
now automatically create it before creating the file.

This change makes the Storage class more robust and user-friendly. Users
will no longer need to manually create directories before running the
jar file.

jerome-neo committed:
Fix find function

jerome-neo committed:
Change JavaFx window settings

jerome-neo committed:
Add Ui.png

jerome-neo committed:
Update user guide

jerome-neo committed:
Update README.md

jerome-neo committed:
Update README.md

jerome-neo committed:
Update README.md

jerome-neo committed:
Fix save files and loading files.

jerome-neo committed:
Merge branch 'master' of https://github.com/jerome-neo/ip

jerome-neo committed:
Improve parsing of commands

jerome-neo committed:
Minor fix to the instruction.

jerome-neo committed:
Fix invalid input return statement
